[SPARK-26682][SQL] Use taskAttemptID instead of attemptNumber for Hadoop.

rdblue opened the pull request:
## What changes were proposed in this pull request?

Updates the attempt ID used by FileFormatWriter. Tasks in stage attempts use the same task attempt number and could conflict. Using Spark's task attempt ID guarantees that Hadoop TaskAttemptID instances are unique.

## How was this patch tested?

Existing tests. Also validated that we no longer detect this failure case in our logs after deployment.

rdblue committed:
Use taskAttemptID instead of attemptNumber for Hadoop, which is unique.

Attempts between stages used the same attempt number and could conflict.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7300/
Test PASSed.

SparkQA commented:
**[Test build #101504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101504/testReport)** for PR 23608 at commit [`a41f291`](https://github.com/apache/spark/commit/a41f2914cb652760d4ff40bb9ee79907901bfbc9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

rdblue commented:
Retest this please.

SparkQA commented:
**[Test build #101548 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101548/testReport)** for PR 23608 at commit [`a41f291`](https://github.com/apache/spark/commit/a41f2914cb652760d4ff40bb9ee79907901bfbc9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7327/
Test PASSed.

vanzin commented:
Is this really needed? `executeTask` seems to be doing the right thing:

```
    val taskId = new TaskID(jobId, TaskType.MAP, sparkPartitionId)
    val taskAttemptId = new TaskAttemptID(taskId, sparkAttemptNumber)
```

Seems like this could be cleaned up a little if you want, but I don't see that old problem happening in this code.

SparkQA commented:
**[Test build #101548 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101548/testReport)** for PR 23608 at commit [`a41f291`](https://github.com/apache/spark/commit/a41f2914cb652760d4ff40bb9ee79907901bfbc9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101548/
Test PASSed.

rdblue commented:
@vanzin, because the attemptNumber resets to 0 for every stage attempt, the Hadoop TaskAttemptID is not unique. Code that assumes that will be unique (like our committer) can produce incorrect results. That's why the `sparkAttemptNumber` that you pointed out should be set to Spark's unique attempt ID instead of the task's attempt number.

vanzin commented:
My bad, misread that code and somehow thought that `taskId` was coming from `TaskContext`. That's what you're doing. D'oh.

I'm a (very) little concerned about the truncation of the task ID, since I was asked to avoid that in the other patch. I really don't think it's a problem, and also don't see how you'd avoid it here, so it looks good, I guess.

rdblue commented:
I think that calling `toInt` should be fine. It is extremely unlikely that taskAttemptID will be larger than 2 billion and it would be fine for this purpose if it were truncated. It would still be unique enough for deduplicating across attempts of the same stage.

The only case I can think of where this would be a problem is if the Hadoop TaskAttemptID can't handle a negative attempt number. I'll update this to `taskContext.taskAttemptId().toInt & Integer.MAX_VALUE` so that this wraps around to 0.

rdblue committed:
Ensure FileFormatWriter sparkAttemptNumber is positive.

This bitwise-ands the unique attempt ID with Integer.MAX_VALUE to ensure
the number is positive, instead of using absolute value. Bitwise-and
ensures that truncation wraps around to 0. Absolute value causes
truncated values to repeat recent IDs:

  (Integer.MAX_VALUE.toLong - 1L).toInt.abs => 2147483646
  (Integer.MAX_VALUE.toLong + 3L).toInt.abs => 2147483646

SparkQA commented:
**[Test build #101601 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101601/testReport)** for PR 23608 at commit [`cbe7eae`](https://github.com/apache/spark/commit/cbe7eaed345156249d81ea9e05e71ce5c6c277a7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7364/
Test PASSed.

dongjoon-hyun commented:
cc @cloud-fan .

zsxwing commented:
Right now `taskAttemptID` is also a partition id. It may be helpful in some case. How about adding the stage attempt number into the path in `org.apache.spark.internal.io.HadoopMapReduceCommitProtocol` instead?

rdblue commented:
@zsxwing, changing Spark's behavior in HadoopMapReduceCommitProtocol isn't related to the problem here. If you think that is broken and will collide as well, please open an issue for it.

The problem is that some Hadoop committer implementations expect the TaskAttemptID to be unique because stages are re-attempted in separate Hadoop MR jobs. Spark should ensure that the TaskAttemptIDs that Spark creates are not reused.

gatorsmile commented:
Which Hadoop committer implementation?

SparkQA commented:
**[Test build #101601 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101601/testReport)** for PR 23608 at commit [`cbe7eae`](https://github.com/apache/spark/commit/cbe7eaed345156249d81ea9e05e71ce5c6c277a7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101601/
Test PASSed.

rdblue commented:
@gatorsmile, the main point is that Spark currently violates the contract for Hadoop OutputCommitter.

A commonly-used committer that makes this assumption is FileOutputCommitter: https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/FileOutputCommitter.java#L266-L269.

cloud-fan commented:
This is consistent with the ds v2 write path, +1

cloud-fan commented:
thanks, merging to master/2.4!

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:
```diff
@@ -170,7 +170,7 @@ object FileFormatWriter extends Logging {
             sparkStageId = taskContext.stageId(),
             sparkPartitionId = taskContext.partitionId(),
-            sparkAttemptNumber = taskContext.attemptNumber(),
+            sparkAttemptNumber = taskContext.taskAttemptId().toInt & Integer.MAX_VALUE,
```
After more than 2 years, I revisited this code path again and realized that this is not the best fix.

The original motivation is still correct: Spark violates the contract of `TaskAttemptID`, as Spark resets task attempt number after stage retry, making `TaskAttemptID` not unique.

The root cause is: Spark job has stages while Hadoop job directly has tasks (no DAG). We map Spark stage id to Hadoop job id, which is inaccurate as this doesn't count stage attempt number.

I think a better fix is to generate the Hadoop job id with both stage id and stage attempt number, or generate the Hadoop task attempt number with both Spark task and stage attempt number. However, the current fix also works, as this only decides the intermedia staging directory name, which we don't care as long as it's unique.

I'm leaving this comment just for future references.
